SI-7442 Update bundled Fork/Join pool (JSR166y)

phaller opened the pull request:
- Updates ForkJoinPool and dependent classes to the latest jsr166y revisions:
  
    ForkJoinPool.java:
    Revision 1.185
    Sat Feb 16 20:50:29 2013 UTC (2 months, 2 weeks ago) by jsr166
  
    ForkJoinTask.java:
    Revision 1.100
    Tue Feb 5 17:09:54 2013 UTC (3 months ago) by jsr166
  
    ForkJoinWorkerThread.java:
    Revision 1.73
    Wed Nov 21 19:54:39 2012 UTC (5 months, 2 weeks ago) by dl
- Includes Akka-contributed `sun.misc.Unsafe` detection to support Android.
  See changeset 06d685c1bbd8a0d058ee8a3f374569f8097f2acc
- Adds private `CountedCompleter` class.
  This class is only visible and used in `ForkJoinPool.java`.

Fixes SI-7442.

phaller commented:
Review by @heathermiller, @axel22, @retronym

heathermiller commented:
For context, this was discussed with Doug Lea on the concurrency-interest mailing list a few days ago:
http://markmail.org/message/kz6joibeoyzvu5ij

phaller commented:
Resolving [86c8d8e3ec025285aa9bcad04eb231cd03a99f13/lib/forkjoin.jar]
Error downloading /home/jenkins/.sbt/cache/scala/86c8d8e3ec025285aa9bcad04eb231cd03a99f13/lib/forkjoin.jar: response code: 404
http://repo.typesafe.com/typesafe/scala-sha-bootstrap/org/scala-lang/bootstrap/86c8d8e3ec025285aa9bcad04eb231cd03a99f13/lib/forkjoin.jar

axel22 commented:
Since this is the newer version of the fork/join framework being copied over the old one, I cannot tell if the changes are correct.
But I certainly agree with the mailing list discussion and Doug's comment there.
So:

LGTM

phaller commented:
Yeah, basically the only way to fully check the correctness of these changes is to do the update yourself and compare the two diffs.

;-)

phaller commented:
I'm looking into the reported binary incompatibilities.

retronym commented on src/forkjoin/scala/concurrent/forkjoin/ForkJoinTask.java:
```diff
@@ -1299,59 +1227,51 @@ public static int getSurplusQueuedTaskCount() {
-            .workQueue.nextLocalTask();
+        Thread t;
+        return ((t = Thread.currentThread()) instanceof ForkJoinWorkerThread) ?
+            ((ForkJoinWorkerThread)t).workQueue.nextLocalTask() :
```
Contrast with:

``` scala
      Thread.currentThread match {
        case fjw: ForkJoinWorkerThread if fjw.getPool eq fj => fjt.fork()
        case _                                              => fj execute fjt
      }
```

(from https://github.com/scala/scala/blob/master/src/library/scala/concurrent/impl/ExecutionContextImpl.scala#L105)

Why does that code need to check for `fjw.getPool eq fj`? Is that check missing here?

axel22 replied:
I believe that the code in the execution context was originally there to use fork only if the current thread is the fork/join worker of the fork/join pool of that execution context -- only in this case we were doing the potentially more efficient `fork` -- otherwise we might get the task scheduled on the wrong pool

Since this is a method on the `ForkJoinTask` and not `ForkJoinPool`, I believe that is the reason there is no such check. Also, this method polls, it does not schedule for execution, so it might not be relevant which pool the polled task comes from.
Just my guess, though.

phaller replied:
Agreed on the reason for the check in `ExecutionContextImpl`.

About the code in `ForkJoinTask`: it's explicitly permitted to create several instances of `ForkJoinPool`, so I trust that the internal `ForkJoinPool` logic ensures that no such check is necessary in this particular method.

From the doc comment it appears that this method is actually designed to really only be relevant to the current thread, regardless of the pool that the thread is part of, since it returns a task that is queued by the current thread (if any):

> Unschedules and returns, without executing, the next task
> queued by the current thread but not yet executed, if the
> current thread is operating in a ForkJoinPool.  This method is
> designed primarily to support extensions, and is unlikely to be
> useful otherwise.

retronym replied:
Sounds okay to me, thanks for the explanation.

phaller commented:
PLS REBUILD/pr-checkin-per-commit@93bca34

scala-jenkins commented:
(kitty-note-to-self: ignore 17666305)
:cat: Roger! Rebuilding pr-checkin-per-commit for 93bca345. :rotating_light:

phaller commented:
PLS REBUILD/pr-rangepos-per-commit@93bca34

scala-jenkins commented:
(kitty-note-to-self: ignore 17713787)
:cat: Roger! Rebuilding pr-rangepos-per-commit for 93bca345. :rotating_light:

phaller commented:
BUILDLOG?

scala-jenkins commented:
(kitty-note-to-self: ignore 17728176)
pr-rangepos-per-commit:
- 93bca345: [error](https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos-per-commit/2092/):  failed: Took 24 s.

pr-checkin-per-commit:
- 93bca345: [error](https://scala-webapps.epfl.ch/jenkins/job/pr-checkin-per-commit/2238/):  failed: Took 20 s.

phaller commented:
PLS REBUILD/pr-checkin-per-commit@93bca34

phaller commented:
PLS REBUILD/pr-rangepos-per-commit@93bca34

scala-jenkins commented:
(kitty-note-to-self: ignore 17734824)
:cat: Roger! Rebuilding pr-checkin-per-commit for 93bca345. :rotating_light:

scala-jenkins commented:
(kitty-note-to-self: ignore 17734834)
:cat: Roger! Rebuilding pr-rangepos-per-commit for 93bca345. :rotating_light:

heathermiller commented:
Conditional upon all of the binary compatibility issues plaguing the build being resolved-- LGTM. 

On an aside, I thought we once decided that internal dependencies like fork join would have slightly more lax compatibility rules? I can't recall the precise details, but maybe @jsuereth can? (I'll look around for where this was discussed)

heathermiller commented:
I'm certain this has been discussed in some capacity, perhaps in a Scala meeting. Though this is the only thing that I can find related to ForkJoin and deprecation: https://github.com/scala/scala/pull/240#commitcomment-1681983

phaller commented:
Here is the latest report on forward binary compatibility issues:
https://scala-webapps.epfl.ch/jenkins/job/pr-checkin-per-commit/2245/console

mima complains that methods like `tryCompensate()Boolean` do not exist in the previous version (the mima messages say "new version", but that's because I presume mima is simply given the two libraries in the other order to check for forward, instead of backward, binary compatibility issues). This `tryCompensate` method has package-private access in the new version, since it is called by a different class in the same package. Making it private does not work:

```
[javac] /Users/phaller/git/scala-2.10.x/src/forkjoin/scala/concurrent/forkjoin/ForkJoinTask.java:1011: tryCompensate() has private access in scala.concurrent.forkjoin.ForkJoinPool
[javac]                         if (p == null || p.tryCompensate())
[javac]                                           ^
```

My current understanding is that the mima settings we use are too strict. Introducing or removing package-private methods in package `scala.concurrent.forkjoin` should not be considered an issue for binary compatibility, since it only affects code within that package. So, only clients who (a) directly use our internal `scala.concurrent.forkjoin` dependency (which is not part of the public API), and (b) put classes or traits into the `scala.concurrent.forkjoin` package are affected. I don't think this is a supported use case, nor should it be supported by our binary compatibility policy.

phaller commented:
Note that the comment that @heathermiller linked to above points out that the corresponding ForkJoin update also changed _public_ methods (that's why @retronym called it "updated wholesale"). In the context of the update of this pull request, the problem is confined to _package-private_ methods in the `forkjoin` package.

retronym commented:
I'm pretty sure this is okay for the internals of FJ; update the whitelist, and I'll try to find precedent.

phaller commented:
OK, good. I updated the binary compatibility whitelists to exclude changed package-private methods in the `forkjoin` package.

I also removed the workarounds I had put in place to avoid similar issues with backward binary compatibility, since they were all necessary because of such changed package-private methods.

phaller commented:
For the record, I maintain a branch that contains all incremental changes as separate commits:
https://github.com/phaller/scala/commits/topic/forkjoin-update-separate-commits

phaller committed:
SI-7442 Update bundled Fork/Join pool (JSR166y)

- Updates ForkJoinPool and dependent classes to the latest jsr166y revisions:

    ForkJoinPool.java:
    Revision 1.185
    Sat Feb 16 20:50:29 2013 UTC (2 months, 2 weeks ago) by jsr166

    ForkJoinTask.java:
    Revision 1.100
    Tue Feb 5 17:09:54 2013 UTC (3 months ago) by jsr166

    ForkJoinWorkerThread.java:
    Revision 1.73
    Wed Nov 21 19:54:39 2012 UTC (5 months, 2 weeks ago) by dl

- Includes Akka-contributed `sun.misc.Unsafe` detection to support Android.
  See changeset 06d685c1bbd8a0d058ee8a3f374569f8097f2acc

- Adds private `CountedCompleter` class.
  This class is only visible and used in `ForkJoinPool.java`.

- Updates desired.sha1 for updated forkjoin.jar.

- Updates binary compatibility whitelists to exclude package-private methods
  in the `forkjoin` package.

- Also fixes SI-7438.

retronym commented:
LGTM. Thanks for getting this done!
